src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala

```
@@ -450,6 +471,34 @@ class GoogleExtensions(val directoryDAO: DirectoryDAO, val accessPolicyDAO: Acce
     }
   }
 
+  private def isConstrainable(resource: Resource, actions: Set[ResourceAction]): Boolean = {
+    resourceTypes.get(resource.resourceTypeName) match {
+      case Some(resourceType) => {
+        for {
+          actionPattern <- resourceType.actionPatterns
+          action <- actions
+        } yield actionPattern.authDomainConstrainable && actionPattern.matches(action)
+      }.contains(true)
+      case None => false // we really shouldn't get here as this means the resource is invalid, but leaving it in for completeness
+    }
+  }
+
+  private def calculateIntersectionGroup(resource: Resource, policy: AccessPolicy): Future[Set[WorkbenchUserId]] = {
+    for {
+      groups <- accessPolicyDAO.loadResourceAuthDomain(resource)
+      authDomain <- Future.traverse(groups)(groupName =>
+        directoryDAO.listFlattenedGroupUsers(groupName))
+      policyMembers <- Future.traverse(policy.members)(groupName => groupName match {
+        case g: WorkbenchGroupName => directoryDAO.listFlattenedGroupUsers(g)
+        case u: WorkbenchUserId => Future.successful(Set(u))
+        case p: PetServiceAccountId => Future(Set(p.userId))
+        case _ => throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.InternalServerError, s"Bad policy found"))
+      })
+    } yield {
+      (authDomain + policyMembers.flatten).reduce(_.intersect(_))
```

if `authDomain + policyMembers.flatten` is empty, `reduce` will throw exception. Will that be an issue? When it's empty, shall we just return empty set or we really want to propagate the error?

---

Is this possible given that `authDomain + policyMembers.flatten` is a `Set[Set[WorkbenchUserId]]`? My understanding is that if there are no users in either the policy or the auth domain, `authDomain + policyMembers.flatten` will still contain the empty set and therefore not be empty. This would result in `reduce` not throwing an exception, but just returning an empty set, which is the desired behavior in this case as there should not be any users included in the intersection group for an empty policy and auth domain.